### DISCUSSION TITLE
Creating Custom Error Pages (404 Redirect)

### DISCUSSION THREAD

**Discussion by latchk3y** [pocketbase/pocketbase#6064] [Q&A] [answered]
This is probably an incredibly basic question, but I'm switching over from using JS to trying to construct what I want in Go, and I'm having difficulty figuring out how exactly I determine if a route is unreachable, and if so, just serve a standard 404 page, rather than the text screen that comes default.

Any help would be appreciated, and my apologies if this was answered somewhere, but I couldn't find it.

**Comment by ganigeorgiev** [comment, marked as answer]
It is a little more complicated because it depends when and for which routes the custom 404 page should be applied because it is possible to receive a 404 response from a valid route (e.g. trying to access a record with unsatisfied View API rule).

I guess one universal solution could be to register a global middleware that will check the returned error and will write your custom 404 page to the response (with exception maybe for the routes starting with `/api/`). For example:

```go
package main

import (
	"errors"
	"log"
	"net/http"
	"strings"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
)

func main() {
	app := pocketbase.New()

	app.OnServe().BindFunc(func(se *core.ServeEvent) error {
		se.Router.BindFunc(func(e *core.RequestEvent) error {
			err := e.Next()
			if err == nil {
				return nil // no error
			}

			apiErr := apis.ToApiError(err)
			if apiErr.Status == http.StatusNotFound && !strings.HasPrefix(e.Request.URL.Path, "/api/") {
				// if you want redirect, call e.Redirect(http.StatusTemporaryRedirect, "https://example.com")
				// for all body write helpers check https://pocketbase.io/docs/go-routing/#writing-response-body
				err = errors.Join(err, e.HTML(http.StatusNotFound, "<h1>404!</h1>"))
			}

			return err
		})

		return se.Next()
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by ganigeorgiev** [reply]
Another option (and maybe better) could be instead of manually excluding `/api/` routes is to check the `Content-Type` request header and return an html response if not `application/json`, aka. something like:

```go
se.Router.BindFunc(func(e *core.RequestEvent) error {
    err := e.Next()
    if err == nil {
        return nil // no error
    }

    apiErr := apis.ToApiError(err)
    if apiErr.Status == http.StatusNotFound &&
        !strings.HasPrefix(e.Request.Header.Get("Content-Type"), "application/json") {
        // if you want redirect, call e.Redirect(http.StatusTemporaryRedirect, "https://example.com")
        // for all body write helpers check https://pocketbase.io/docs/go-routing/#writing-response-body
        err = errors.Join(err, e.HTML(http.StatusNotFound, "<h1>404!</h1>"))
    }

    return err
})
```

**Reply by latchk3y** [reply]
both of these answers work like a charm. it also addresses my follow-up question concerning how i would see the header content for these types of checks as well.

thank you much, exactly what i needed.
